llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -1396,8 +1425,16 @@ static bool runImpl(Module &M, AnalysisGetter &AG, TargetMachine &TM,
     }
   }
 
-  ChangeStatus Change = A.run();
-  return Change == ChangeStatus::CHANGED;
+  bool Changed = A.run() == ChangeStatus::CHANGED;
+
+  // We only update the waves-per-eu attribute at the final stage to avoid
+  // setting it with intermediate values.
+  if (Changed && (LTOPhase == ThinOrFullLTOPhase::None ||
+                  LTOPhase == ThinOrFullLTOPhase::FullLTOPostLink ||
+                  LTOPhase == ThinOrFullLTOPhase::ThinLTOPostLink))
```

Comment:
This doesn't work. You can't not set it as later Changed might be false, and you never validate/fixup the result. I would suggest to ignore Changed and simply update the WavesPerEu.